src/graph/scheduler/AsyncMsgNotifyBasedScheduler.cpp

```diff
@@ -225,17 +239,94 @@ void AsyncMsgNotifyBasedScheduler::notifyError(std::vector<folly::Promise<Status
 }
 
 folly::Future<Status> AsyncMsgNotifyBasedScheduler::execute(Executor* executor) const {
-  memory::MemoryCheckGuard guard1;
   auto status = executor->open();
   if (!status.ok()) {
     return executor->error(std::move(status));
   }
-  return executor->execute().thenValue([executor](Status s) {
-    memory::MemoryCheckGuard guard2;
-    NG_RETURN_IF_ERROR(s);
-    return executor->close();
+
+  auto exeStatus = runExecute(executor);
+
+  return std::move(exeStatus).thenValue([this, executor](Status s) {
+    if (!s.ok()) {
+      DLOG(INFO) << formatPrettyId(executor) << " failed with: " << s.toString();
+      setFailStatus(s);
+      removeExecuting(executor);
+      return Status::from(s);
+    }
+    auto ret = executor->close();
+    removeExecuting(executor);
+    return ret;
   });
 }
 
+folly::Future<Status> AsyncMsgNotifyBasedScheduler::runExecute(Executor* executor) const {
+  // catch Executor::execute here, upward call stack should only get Status, no exceptions.
+  try {
+    addExecuting(executor);
+    if (hasFailStatus()) return failedStatus_.value();
+    folly::Future<Status> status = Status::OK();
+    {
+      memory::MemoryCheckGuard guard;
+      status = executor->execute();
```

Comment:
plz NOTE: the `execute` is an async function, your memory check guard may be invalid when the tasks really run.

Reply:
> plz NOTE: the `execute` is an async function, your memory check guard may be invalid when the tasks really run.

This MemoryCheckGuard is only used by some Executor return a already satisfied future, like DedupExecutor::execute, Async future's MemoryCheckGuard is not here, they are declare inside, for example in StorageClientBase-inl.h:

```
  return folly::collectAll(respFutures)
      .deferValue([this, requests = std::move(requests), totalLatencies, hosts](
                      std::vector<folly::Try<StatusOr<Response>>>&& resps) {
        // throw in MemoryCheckGuard verified
        memory::MemoryCheckGuard guard;
        StorageRpcResponse<Response> rpcRes
```